### PR TITLE
use predefined resolutions for thumbnail generation

### LINKED ISSUE #7: Prevent generation of large amount of files.
By incrementing the requested resolution in small steps a malicous actor could flood the service with thumbnails and thus fill the storage with a large amount of files.

### PR DESCRIPTION
**Pull request by C0rby**
Use predefined resolutions to prevent flooding with a large amount of thumbnails.
Each file can only have as many thumbnails as there are defined resolutions.

Closes #7

### CODE REVIEW

File: pkg/thumbnails/resolution/resolutions.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	}
+
+	isLandscape := width > height
+	givenLen := int(math.Max(float64(width), float64(height)))
+
+	// Initialize with the first resolution
+	var match Resolution
+	minDiff := math.MaxInt32
+
+	for i := 1; i < len(r); i++ {
```
Comment (refs) [inline pkg/thumbnails/resolution/resolutions.go]: Could we stick to iterating using `range`?

File: pkg/thumbnails/resolution/resolutions.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
+package resolution
+
+import (
+	"fmt"
+	"math"
+	"sort"
+)
+
+// Init creates an instance of Resolutions from resolution strings.
+func Init(rStrs []string) (Resolutions, error) {
```
Comment (refs) [inline pkg/thumbnails/resolution/resolutions.go]: call it `New`. It's more aligned with the style around our codebase ^^

File: pkg/thumbnails/storage/storage.go [hunk 3]
```diff
@@ -1,11 +1,12 @@
 package storage
 
+import "github.com/owncloud/ocis-thumbnails/pkg/thumbnails/resolution"
+
 // Context combines different attributes needed for storage operations.
 type Context struct {
```
Comment (refs) [inline pkg/thumbnails/storage/storage.go]: Have you give it a thought about using golang's context api? :D

File: pkg/thumbnails/thumbnails.go [hunk 4]
```diff
@@ -69,16 +69,15 @@ func (s SimpleManager) GetStored(ctx Context) []byte {
 }
 
 func (s SimpleManager) generate(ctx Context, img image.Image) image.Image {
-	thumbnail := resize.Thumbnail(uint(ctx.Width), uint(ctx.Height), img, resize.Lanczos2)
+	thumbnail := resize.Thumbnail(uint(ctx.Resolution.Width), uint(ctx.Resolution.Height), img, resize.Lanczos2)
```
Comment (IljaN) [inline pkg/thumbnails/thumbnails.go]: Is it good practice to pass domain-values inside context?

File: pkg/thumbnails/storage/storage.go (hunk 3 above)
Comment (C0rby) [reply pkg/thumbnails/storage/storage.go]: Yes, I did but isn't golangs context api a bit overkill for our use case?
And also a bit too generic for a method api?

File: pkg/thumbnails/thumbnails.go (hunk 4 above)
Comment (C0rby) [reply pkg/thumbnails/thumbnails.go]: Currently it is a custom context defined in this package.
Probably context is a bad name but I couldn't think of a fitting one.

It's an object that combines different values needed to fullfill a thumbnail request.
Maybe I should rename it to Request? :D

Comment (IljaN) [reply pkg/thumbnails/thumbnails.go]: Ah :bulb:  If it is a custom struct then it makes sense to rename it to something else to reduce confusion with golang contexts.

File: pkg/thumbnails/storage/storage.go (hunk 3 above)
Comment (refs) [reply pkg/thumbnails/storage/storage.go]: Context is not overkill, it's the simplest package out there :P

Also:

>  Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx

with emphasis on ctx as first argument.

I think what you're referring to as context, as you said, can be something else.

The pros of using `context.Context` is once you adopt the idiom, goroutines get way, way easier; if you intend using goroutines here anyway.

I think renaming it to something else would be less confusing :)

File: pkg/thumbnails/resolutions/resolution.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
+package resolutions
```
Comment (refs) [inline pkg/thumbnails/resolutions/resolution.go]: package names should be singular

**Pull request merged**